### PR TITLE
Add support for 2024 G14 (GA403UI)

### LINKED ISSUE #21: Support for Asus Zephyrus G14 2024 (GA403UI)
I would like to help out bringing support for the 2024 G14 model (GA403UI).
I have little experience with Rust, but looking at the code, I think I'll manage.

My first main focus would be to add control for the newly added ledstrip on the back (which replaces the Animatrix if I'm not mistaken).

Looking at the rog-anime code, it seems like you communicate over USB using a specific Asus created protocol?
My hardware coding experiences are pretty rudimentary. Would it be possible to point me in some direction on how one could tackle this problem?
Maybe, the "Anime protocol" can even partly  be re-used or act as a starting point.

If anyone else has already been looking into this, please let me know :) 

Have a nice day!

### PR DESCRIPTION
**Pull request by jschoubben**
The 2024 G14 has the Anime Matrix Display replaced with a Slash Led Display. This pull-requests includes support for this new device.

Done
[x] [asusd] Add ZBus org.asuslinux.Slash
[x] [asusctl] Add enable functionality
[x] [asusctl] Support brightness changes
[x] [asusctl] Support "interval" setting
[x] [asusctl] Supoort changing the Animation (SlashMode)

Todo
[ ] [asusd] Add support to run a specific animation once on "events"
[ ] [rog-control-center] Add a new screen to configure the Slash

ps. I wasn't able to commit/push without the `--no-verify` flag enabled. the `cargo test --all` command fails on 2 tests regarding the Aura device so I'm not sure how you would like me to approach this.

Looking forward to any feedback!

resolves #21

### CODE REVIEW

**Comment by jschoubben** [comment]
#23

**Comment by jschoubben** [comment]
Resolves #21

**Comment by flukejones** [comment]
Lucky I checked github. It doesn't email me about PR or issues.

Are you able to add a simulator for this please? The existing Anime one should be a good starting point.

File: asusd/src/config.rs [hunk 1]
```diff
@@ -90,13 +90,13 @@ impl StdConfig for Config {
...
     }
 
-    fn config_dir() -> std::path::PathBuf {
-        std::path::PathBuf::from(crate::CONFIG_PATH_BASE)
-    }
-
     fn file_name(&self) -> String {
         CONFIG_FILE.to_owned()
     }
+
```
Comment (flukejones) [inline asusd/src/config.rs:96]: Please don't do spurious changes like this, it adds noise to the meat of the PR.

File: asusd/src/ctrl_anime/config.rs [hunk 2]
```diff
@@ -150,13 +150,13 @@ impl StdConfig for AnimeConfig {
...
     }
 
-    fn config_dir() -> std::path::PathBuf {
-        std::path::PathBuf::from(crate::CONFIG_PATH_BASE)
-    }
-
     fn file_name(&self) -> String {
         CONFIG_FILE.to_owned()
     }
+
```
Comment (flukejones) [inline asusd/src/ctrl_anime/config.rs:156]: Same again, spurious change

File: asusd/src/ctrl_aura/manager.rs [hunk 3]
```diff
@@ -5,7 +5,6 @@
 // - If udev sees device removed then remove the zbus path
 
 use std::collections::HashSet;
-use std::time::Duration;
```
Comment (flukejones) [inline asusd/src/ctrl_aura/manager.rs:8]: Unrequired change. These things should be in separate PR or commits.

File: asusd/src/ctrl_slash/mod.rs [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+        }
+        Ok(())
+    }
+}
+
+pub struct CtrlSlash {
+    // node: HidRaw,
+    node: Node,
+    config: SlashConfig,
+    // slash_type: SlashType,
```
Comment (flukejones) [inline asusd/src/ctrl_slash/mod.rs:36]: There is the G16 version, if you can please add handling for it and we will assume it is the same length for now.

File: asusd/src/ctrl_slash/mod.rs [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    // // set to force thread to exit
```
Comment (flukejones) [inline asusd/src/ctrl_slash/mod.rs:37]: Remove all commented sections. YAGNI principle.

File: asusd/src/ctrl_slash/mod.rs [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+
+        let slash_type = get_slash_type()?;
+        if slash_type == SlashType::Unknown  {
+            return Err(RogError::Slash(SlashError::NoDevice));
+        }
+
+        let ctrl = CtrlSlash {
+            node,
+            config,
+            // slash_type,
```
Comment (flukejones) [inline asusd/src/ctrl_slash/mod.rs:64]: Remove all commented sections. YAGNI principle.

File: asusd/src/lib.rs [hunk 7]
```diff
@@ -130,7 +132,7 @@ pub fn print_board_info() {
 }
 
 pub trait Reloadable {
-    fn reload(&mut self) -> impl std::future::Future<Output = Result<(), RogError>> + Send;
```
Comment (flukejones) [inline asusd/src/lib.rs:133]: Please avoid these sorts of changes if they are not directly related to the task on hand.

**Review by flukejones — COMMENTED** [review]
This looks pretty good. But there is a lot of noise in the PR due to changes unrelated. I suspect some of this is due to you doing a merge of main in to your work.

Can you please do a rebase on to main to try and clean this up.

**Comment by jschoubben** [comment]
I'll get to it as soon as I can.

I wasn't sure about your views/opinions regarding PR's (it's clear now), I'll try to make it as pristine as possible.
I use RustRover which makes a lot of "suggestions" to optimize/clean/improve your codebase. 

Also, after merging main, rog-control-center started crashing due to the newly add tray functionality. 

Thanks for the review!

**Comment by flukejones** [comment]
It's not a super major, but if you rebase it should help a little - and don't worry too much about the extraneous changes.

Yeah... That tray has become a problem - I'll have to either revert back to the old one or make changes upstream.

**Comment by flukejones** [comment]
Hey mate I might merge this now so I can prevent more conflicts.

**Comment by jschoubben** [comment]
I'm sorry, we're having a tight deadline at work and I didn't find the time yet to resolve your comments.
It's still on my todo-list, I'll get back to it as soon as I can.

Sorry for the delay!
